Activating a plugin should fail when the plugin has missing dependencies

mattab opened the pull request:
Bug: when using `plugin:activate` console command: without this patch it looks like the plugin:activate worked but it didn't

mattab committed:
Activating a plugin should fail when the plugin has missing dependencies


For example via plugin:activate console command.
without this patch it looks like the plugin:activate worked but it didn't

tsteur commented on core/Plugin/Manager.php:
```diff
@@ -1374,4 +1377,24 @@ public function loadPluginTranslations()
+     * @param $plugin Plugin
+     * @throws \Exception
+     */
+    private function throwIfPluginMissingDependencies($plugin)
```
Can you maybe move this method in another class and make it public so it can be tested? the string formatting part might be also useful for plugins/CorePluginsAdmin/templates/macros.twig and https://github.com/piwik/piwik/blob/2.16.2/core/Plugin/Manager.php#L831 where it should be actually deactivated the plugin even before we try to install I think

mattab committed:
Reuse the new method in twig

mattab committed:
Slight refactor for clarity

mattab committed:
Better code org

mattab committed:
Add integration test

mattab commented:
Added the text to the notification: ![reuse](https://cloud.githubusercontent.com/assets/466765/17720336/2b18bb80-6475-11e6-8e60-59b691763eea.png)

The twig code now reuses the PHP logic

Added test for the new method

mattab committed:
Spacing

mattab committed:
fix tests

tsteur commented on core/Plugin/Dependency.php:
```diff
@@ -77,6 +77,25 @@ public function setPiwikVersion($piwikVersion)
+
+        $causedBy = array();
+        foreach ($missingDependencies as $dependency) {
+            $causedBy[] = strtoupper($dependency['requirement']) . ' ' . $dependency['causedBy'];
```
I think this should be rather capitalize instead of strtoupper. There will be also plugin names etc. For PHP we have in the Marketplace a special if to convert it to upper case

tsteur commented on plugins/CorePluginsAdmin/Commands/ActivatePlugin.php:
```diff
@@ -37,6 +37,11 @@ protected function execute(InputInterface $input, OutputInterface $output)
             return;
         }
 
+        if ($dependencies = $pluginManager->getPluginMissingDependenciesAsString($plugin)) {
```
I wonder if we should directly use the PluginManager here to get the loaded plugin, and directly use dependency class. If we otherwise move everything that loads a plugin or something into Plugin\Manager we end up having hundreds of methods there. Also it adds a formatting method is in `Dependency` so that class is actually doing many things. If I may suggest an improvement I would say to rather add a method `getMissingDependencies` to the `Plugin` class and then do `Plugin\Manager->loadPlugin(pluginName)->getMissingDependencies()`.

Ideally we would have a class `MissingDependencies` that extends ArrayObject and is therefore iterable and also cares about the formatting. I'm saying this because we will soon move this part into composer but we can also do this change later and refactor it again soon.

mattab committed:
use ucfirst and not strtoupper

mattab committed:
test: use ucfirst and not strtoupper

mattab committed:
Simplify code as per code review

mattab committed:
Simplify code as per code review

mattab committed:
EOL LF

mattab committed:
Merge branch '3.x-dev' into missingdeps_activate

Conflicts:
	libs/bower_components/jquery.scrollTo/README.md
	libs/bower_components/jquery.scrollTo/jquery.scrollTo.min.js
	libs/bower_components/jquery.scrollTo/scrollTo.jquery.json
	plugins/CorePluginsAdmin/lang/en.json
	plugins/CorePluginsAdmin/templates/macros.twig

mattab committed:
Remove test

mattab commented:
Will re-create the PR clean, as I included submodules by mistake
